samples/_emscripten/_audio/EmscriptenAudio/Voice.cpp

```diff
@@ -0,0 +1,345 @@
+
+VoiceRef Voice::create( const CallbackProcessorFn &callbackFn, const Options &options )
+{
+	VoiceRef result( new VoiceCallbackProcessor( callbackFn, options ) );
+	MixerImpl::get()->addVoice( result, options );
+
+	return result;
+}
+
+bool Voice::isDoneLoading(){
+	return mDoneParsing;
+}
+
+void Voice::create( const SourceFileRef &sourceFile, std::function<void(VoiceSamplePlayerNodeRef)> cb, const Options &options){
+	new VoiceSamplePlayerNode( sourceFile, options , cb );
+}
+VoiceSamplePlayerNodeRef Voice::create( const SourceFileRef &sourceFile, const Options &options )
+{
+	VoiceSamplePlayerNodeRef result( new VoiceSamplePlayerNode( sourceFile, options ) );
+	MixerImpl::get()->addVoice( result, options );
+
+	return result;
+}
+
+Voice::~Voice()
+{
+	MixerImpl::get()->removeVoice( mBusId );
+}
+
+void Voice::clearBufferCache()
+{
+	MixerImpl::get()->clearBufferCache();
+}
+
+float Voice::getVolume() const
+{
+	return MixerImpl::get()->getBusVolume( mBusId );
+}
+
+float Voice::getPan() const
+{
+	return MixerImpl::get()->getBusPan( mBusId );
+}
+
+void Voice::setVolume( float volume )
+{
+	MixerImpl::get()->setBusVolume( mBusId, volume );
+}
+
+void Voice::setPan( float pan )
+{
+	MixerImpl::get()->setBusPan( mBusId, pan );
+}
+
+void Voice::start()
+{
+	getInputNode()->enable();
+}
+
+void Voice::pause()
+{
+	getInputNode()->disable();
+}
+
+void Voice::stop()
+{
+	getInputNode()->disable();
+}
+
+bool Voice::isPlaying() const
+{
+	return getInputNode()->isEnabled();
+}
+
+NodeRef Voice::getOutputNode() const
+{
+	return MixerImpl::get()->getOutputNode( mBusId );
+}
+
+// ----------------------------------------------------------------------------------------------------
+// VoiceSamplePlayerNode
+// ----------------------------------------------------------------------------------------------------
+
+
+VoiceSamplePlayerNode::VoiceSamplePlayerNode( const SourceFileRef &sourceFile, const Options &options, std::function<void(VoiceSamplePlayerNodeRef)> cb )
+	: Voice()
+{
+
+
+	#ifndef CINDER_EMSCRIPTEN
+		size_t requiredSampleRate = audio::master()->getSampleRate();
+		SourceFileRef sf = requiredSampleRate == sourceFile->getSampleRate() ? sourceFile : sourceFile->cloneWithSampleRate( requiredSampleRate );
+
+		if( sf->getNumFrames() <= options.getMaxFramesForBufferPlayback() ) 
+		{
+			BufferRef buffer = MixerImpl::get()->loadBuffer( sf );
+			mNode = Context::master()->makeNode( new BufferPlayerNode( buffer ) );
+		} 
+		else {
+			mNode = Context::master()->makeNode( new FilePlayerNode( sf ) );
+		}
+	#else
+
+
+		auto loader = (WebAudioLoader*)sourceFile.get();
+
+		loader->init( [=]()->void
+		{
+			BufferRef buffer = MixerImpl::get()->loadBuffer( sourceFile );
+			BufferRef audio = loader->getAudioData();
+			mNode = Context::master()->makeNode( new BufferPlayerNode( audio ) );
+			mDoneParsing = true;
+
+			VoiceSamplePlayerNodeRef result( this );
```

Comment:
This is incorrect usage of `shared_ptr`s here. You can't safely create a shared_ptr from the shared object's constructor (even if you use `shared_from_this()`), you should instead use the `create()` methods ([here](https://github.com/cinder/Cinder/blob/master/src/cinder/audio/Voice.cpp#L197-L203)) to construct the `shared_ptr` and add it to the mixer.

Reply:
hmmm - I get what you're saying but I don't know if there's a good way to avoid this; I spent some time this afternoon trying to re-write things to build a `shared_ptr` in a more normal fashion, for example 

```c++
VoiceSamplePlayerNodeRef result = VoiceSamplePlayerNodeRef(new VoiceSampleNode) 
``` 

and every time I try to add the `VoiceSamplePlayerNodeRef` to the mixer, I get a memory error saying I accessed something out of bounds(unfortunately the error isn't too helpful beyond that).

I will take a look again later.